File: src/main/java/com/example/finsentinel/service/trading/uta/PaperBroker.java
package com.example.finsentinel.service.trading.uta;

import com.example.finsentinel.service.trading.engine.AccountInfo;
import com.example.finsentinel.service.trading.engine.OrderRequest;
import com.example.finsentinel.service.trading.engine.OrderResult;
import com.example.finsentinel.service.trading.engine.PaperTradingEngine;
import com.example.finsentinel.service.trading.engine.PositionInfo;

import java.util.EnumSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;

/**
 * IBroker adapter for the in-memory paper trading engine.
 * Supports all security types since paper trading simulates everything.
 * Not a Spring bean — created by BrokerRegistry.
 */
public class PaperBroker implements IBroker {

    private final PaperTradingEngine engine;

    public PaperBroker(PaperTradingEngine engine) {
        this.engine = Objects.requireNonNull(engine, "PaperTradingEngine must not be null");
    }

    @Override
    public String brokerId() {
        return "paper";
    }

    @Override
    public String displayName() {
        return "Paper Trading (Simulated)";
    }

    @Override
    public Set<SecurityType> supportedSecurityTypes() {
        return EnumSet.allOf(SecurityType.class);
    }

    @Override
    public Set<BrokerCapability> capabilities() {
        return EnumSet.of(BrokerCapability.SPOT_TRADING, BrokerCapability.MARKET_DATA);
    }

    @Override
    public OrderResult placeOrder(Contract contract, OrderRequest request) {
        var nativeRequest = new OrderRequest(
                contract.toEngineSymbol(), request.side(), request.type(),
                request.qty(), request.notional(), request.price(),
                request.stopPrice(), request.timeInForce(), request.reduceOnly());
        return engine.placeOrder(nativeRequest);
    }

    @Override
    public List<PositionInfo> getPositions() {
        return engine.getPositions();
    }

    @Override
    public List<OrderResult> getOrders() {
        return engine.getOrders();
    }

    @Override
    public AccountInfo getAccount() {
        return engine.getAccount();
    }

    @Override
    public boolean cancelOrder(String orderId) {
        return engine.cancelOrder(orderId);
    }

    /** Exposes the underlying engine for wallet state sync. */
    public PaperTradingEngine engine() {
        return engine;
    }
}

File: apps/api/src/trading/brokers/paper.broker.ts
import {
  SecurityType,
  BrokerCapability,
  isCrypto,
  Contract,
} from '@finsentinel/shared';
import type { IBroker } from '../interfaces/broker';
import type { PaperTradingEngine } from '../engines/paper-trading.engine';
import type {
  OrderRequest,
  OrderResult,
  PositionInfo,
  AccountInfo,
  MarketClock,
} from '../interfaces/types';

/**
 * All SecurityType values as a Set.
 * Paper broker supports every instrument type.
 */
const ALL_SECURITY_TYPES: Set<SecurityType> = new Set([
  SecurityType.STOCK,
  SecurityType.OPTION,
  SecurityType.FUTURE,
  SecurityType.CRYPTO,
  SecurityType.PERP,
  SecurityType.FOREX,
]);

/**
 * PaperBroker capabilities — spot trading and market data only.
 */
const PAPER_CAPABILITIES: Set<BrokerCapability> = new Set([
  BrokerCapability.SPOT_TRADING,
  BrokerCapability.MARKET_DATA,
]);

/**
 * PaperBroker — wraps PaperTradingEngine with Contract-aware symbol conversion.
 *
 * NOT an Injectable. Created by BrokerRegistry at runtime.
 *
 * Paper broker adapter:
 * - Converts crypto/perp contracts to Polygon format: "X:{symbol}{currency}"
 * - Delegates all operations to the underlying PaperTradingEngine
 */
export class PaperBroker implements IBroker {
  constructor(private readonly _engine: PaperTradingEngine) {}

  // ── Identity ────────────────────────────────────────────────────────────

  brokerId(): string {
    return 'paper';
  }

  displayName(): string {
    return 'Paper Trading (Simulated)';
  }

  // ── Capability checks ──────────────────────────────────────────────────

  supportedSecurityTypes(): Set<SecurityType> {
    return ALL_SECURITY_TYPES;
  }

  capabilities(): Set<BrokerCapability> {
    return PAPER_CAPABILITIES;
  }

  canHandle(contract: Contract): boolean {
    return this.supportedSecurityTypes().has(contract.secType);
  }

  // ── Order execution ────────────────────────────────────────────────────

  async placeOrder(
    contract: Contract,
    request: OrderRequest,
  ): Promise<OrderResult> {
    // Convert contract to Polygon-native symbol for crypto: "X:{symbol}{currency}"
    const nativeSymbol = isCrypto(contract.secType)
      ? `X:${contract.symbol}${contract.currency}`
      : contract.symbol;

    const adapted: OrderRequest = {
      ...request,
      symbol: nativeSymbol,
    };

    return this._engine.placeOrder(adapted);
  }

  // ── Delegated operations ───────────────────────────────────────────────

  async getPositions(): Promise<PositionInfo[]> {
    return this._engine.getPositions();
  }

  async getOrders(): Promise<OrderResult[]> {
    return this._engine.getOrders();
  }

  async getAccount(): Promise<AccountInfo> {
    return this._engine.getAccount();
  }

  async cancelOrder(orderId: string): Promise<boolean> {
    return this._engine.cancelOrder(orderId);
  }

  async syncOrders(): Promise<OrderResult[]> {
    return this._engine.syncOrders();
  }

  async getMarketClock(): Promise<MarketClock> {
    return this._engine.getMarketClock();
  }

  async searchContracts(_query: string): Promise<Contract[]> {
    return [];
  }

  // ── Engine access ──────────────────────────────────────────────────────

  /** Expose the underlying engine for wallet state management. */
  engine(): PaperTradingEngine {
    return this._engine;
  }
}
